pygama/dsp/_processors/multi_t_filter.py

```diff
@@ -0,0 +1,169 @@
+    Returns
+    -------
+    max_list, min_list : array-like, array-like
+        Arrays of fixed length (padded with nans) that hold the indices of
+        the identified local maxima and minima
+    n_max, n_min: scalar, scalar 
+        The number of maxima and minima found in a waveform
+    flag: scalar
+        Returns 0 if there is only one maximum and it is a simple waveform, 
+        returns 1 if there are multiple peaks in a waveform
+        
+    """
+
+    # prepare output 
+    # padded with nans or zeros depending on what we want
+    max_list[:]= np.nan
+    min_list[:] = np.nan
+    n_max[0] = 0
+    n_min[0] = 0
+    flag[0] = 0
+    
+
+    # sanity checks
+    data, delta2 = np.asarray(data), np.asarray(delta)
+    if delta2.ndim != 0:
+        raise Exception("Input argument delta must be a scalar")
+        return np.array([]), np.array([])
+    if delta <= 0:
+        raise Exception("Input argument delta must be positive")
+        return np.array([]), np.array([])
+
+    # now loop over data
+    imax, imin = 0, 0
+    find_max = True
+    for i in range(len(data)):
+
+        if data[i] > data[imax]: imax = i
+        if data[i] < data[imin]: imin = i
+
+        if find_max:
+            # if the sample is less than the current max by more than delta,
+            # declare the previous one a maximum, then set this as the new "min"
+            if data[i] < data[imax] - delta and int(n_max[0]) < int(len(max_list)):
+                max_list[int(n_max[0])]=imax
+                n_max[0] += 1
+                imin = i
+                find_max = False
+        else:
+            # if the sample is more than the current min by more than delta,
+            # declare the previous one a minimum, then set this as the new "max"
+            if data[i] > data[imin] + delta and int(n_min[0]) < int(len(min_list)):
+                min_list[int(n_min[0])] = imin
+                n_min[0] += 1
+                imax = i
+                find_max = True
+    if n_max[0] != 0:
+        flag[0] = 1
+
+
+
+@njit
+def njit_time_point_thresh(wf_in, threshold, tp_max, tp_out):
+    """
+    Find the last timepoint before tp_max that wf_in crosses a threshold
+     wf_in: input waveform
+     threshold: threshold to search for
+     tp_max: time of a maximum of a waveform that the search starts at
+     tp_out: final time that waveform is less than threshold
+    """
+    for i in range(tp_max, 0, -1):
+        if(wf_in[i]>threshold and wf_in[i-1]<threshold):
+            tp_out = i
+            return tp_out
+    tp_out = 0
+    return tp_out
+    
+
+
+@guvectorize(["void(float32[:], float32[:],float32[:], float32, float32[:])",
+              "void(float64[:], float64[:], float64[:], float64, float64[:])",
+              "void(int32[:], int32[:], int32[:], int32, int32[:])",
+              "void(int64[:], int64[:], int64[:], int64, int64[:])"],
+             "(n),(m),(),()->(m)", nopython=True, cache=True)
```

Comment:
As I said in my longer comment, we may want to write this using `forceobj=True` instead of nopython mode so that we can call a series of guvectorized functions